akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.delivery
+
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.TimerScheduler
+import akka.actor.typed.scaladsl.LoggerOps
+
+// FIXME Scaladoc describes how it works, internally. Rewrite for end user and keep internals as impl notes.
+
+/**
+ * The `ProducerController` sends `RequestNext` to the actual producer, which is then allowed to send
+ * one message to the `ProducerController`. The `RequestMessage` message is defined via a factory
+ * function so that the producer can decide what type to use. The producer and `ProducerController`
+ * are supposed to be local so that these messages are fast and not lost.
+ *
+ * The `ProducerController` sends the first message to the `ConsumerController` without waiting for
+ * a `Request` from the `ConsumerController`. The main reason for this is that when used with
+ * Cluster Sharding the first message will typically create the `ConsumerController`. It's
+ * also a way to connect the ProducerController and ConsumerController in a dynamic way, for
+ * example when the ProducerController is replaced.
+ *
+ * When the first message is received by the `ConsumerController` it sends back the initial `Request`.
+ *
+ * Apart from the first message the `ProducerController` will not send more messages than requested
+ * by the `ConsumerController`.
+ *
+ * When there is demand from the consumer side the `ProducerController` sends `RequestNext` to the
+ * actual producer, which is then allowed to send one more message.
+ *
+ * Each message is wrapped by the `ProducerController` in [[ConsumerController.SequencedMessage]] with
+ * a monotonically increasing sequence number without gaps, starting at 1.
+ *
+ * The `Request` message also contains a `confirmedSeqNr` that is the acknowledgement
+ * from the consumer that it has received and processed all messages up to that sequence number.
+ *
+ * The `ConsumerController` will send [[ProducerController.Internal.Resend]] if a lost message is detected
+ * and then the `ProducerController` will resend all messages from that sequence number. The producer keeps
+ * unconfirmed messages in a buffer to be able to resend them. The buffer size is limited
+ * by the request window size.
+ *
+ * The resending is optional, and the `ConsumerController` can be started with `resendLost=false`
+ * to ignore lost messages, and then the `ProducerController` will not buffer unconfirmed messages.
+ * In that mode it provides only flow control but no reliable delivery.
+ */
+object ProducerController {
```

Comment:
Is there a way for the producer to signal that it is done/reached the end of the messages it wants to send etc?

Reply:
no, good idea

Reply:
Changing my mind, better to have that as an application level end message. Because it needs the same delivery mechanism and the application consumer must be able to understand it. When it has been received the application consumer can tear things down.

Reply:
Ok, so that means the CC stopping on consumer termination?

Reply:
yes, that is implemented

on producer side I think the producer would have to wait for ack corresponding to that last message before stopping the PC